examples/multi_modal/loop.py

```diff
@@ -0,0 +1,333 @@
+        """
+        Call the eval share for training
+        Args:
+            batch: tensor
+        Returns:
+            dict with loss, outputs and ground_truth
+        """
+        return self._shared_eval_step(batch, "train")
+
+    def validation_step(self, batch: Dict[str, torch.Tensor], batch_idx: int) -> Dict:
+        """
+        Call the eval share for validation
+        Args:
+            batch:
+            batch_idx:
+        Returns:
+            dict with loss, outputs and ground_truth
+        """
+        return self._shared_eval_step(batch, "val")
+
+    def test_step(self, batch: Dict[str, torch.Tensor], batch_idx: int) -> Dict:
+        """
+        Call the eval share for test
+        Args:
+            batch:
+            batch_idx:
+        Returns:
+            dict with loss, outputs and ground_truth
+        """
+        ret = self._shared_eval_step(batch, "test")
+        self.pred_dict_list.append(ret)
+        return ret
+
+    def _shared_eval_step(self, batch: Dict[str, torch.Tensor], mode: str) -> Dict:
+        """Calculate the desired metrics.
+
+        Args:
+            batch: tensor
+            mode: train, test or val
+        Returns:
+            dict with loss, outputs and ground_truth
+
+        """
+        ids = batch["ID"]
+        atts = batch["Att"]
+        ground_truth = batch["GT"]
+        numerical_id = batch["NID"]
+        img = batch["IMG"]
+        out = self.forward(ids, atts, img)
+        if mode == "train":
+            loss = self.criterion(out, ground_truth)
+            output = self.train_metrics(out, ground_truth)
+            self.log_dict(output)
+            self.train_metrics.update(out, ground_truth)
+        elif mode == "val":
+            loss = self.criterion(out, ground_truth)
+            output = self.val_metrics(out, ground_truth)
+            self.val_metrics.update(out, ground_truth)
+        elif mode == "test":
+            loss = self.criterion(out, ground_truth)
+            output = self.test_metrics(out, ground_truth)
+            self.test_metrics.update(out, ground_truth)
+
+        return {"outputs": out, "loss": loss, "ground_truth": ground_truth, "numerical_id": numerical_id}
+
+    def on_test_end(self) -> None:
+        """Keep the test batches for reporting metrics, like prediction table and classification report."""
+        output_list = []
+        ground_truths_list = []
+        numerical_ids_list = []
+        for batch_item in self.pred_dict_list:
+            output_list.append(batch_item["outputs"])
+            ground_truths_list.append(batch_item["ground_truth"])
+            numerical_ids_list.append(batch_item["numerical_id"])
+        output_tensor = torch.concat(output_list, dim=0)
+        ground_truths_tensor = torch.concat(ground_truths_list, dim=0)
+        numerical_ids_tensor = torch.concat(numerical_ids_list, dim=0)
+        self._sklearn_metrics(output_tensor, ground_truths_tensor, "test", numerical_ids_tensor)
```

Comment:
self.pred_dict_list = [] needs to be reset, otherwise you have a memory leak everytime you call .test

Reply:
But test is only called ones, when training and validation are done. So there will be no multiply epochs or call, correct?

Reply:
done